---
title: "Senior capstone project"
output: html_notebook
---

This is the executable code for my senior capstone project, "Investigating the roles of innate immune cells in intestinal immunity and gut microbiome signaling." 

---
Pre-processing
---

First, the relevant packages are loaded. 

```{r Loading packages}

library(Seurat)
library(patchwork)
library(pathfindR)
library(sqldf)

```

The relevant files from indicated datasets (GSE150050, GSE185224, and GSE125527) are imported from GEO into the R environment. They are then converted to expression matrices. 

GSE150050 contains scRNA-seq data from four tissue locations. This data will be narrowed using SQL on metadata files to isolate colon samples and clustered ('Seurat clustering & ILC phenotype analysis'); ILC expression data will be extracted and subsequently used in the 'Metabolic pathway analysis' section. 

GSE185224 contains primary scRNA-seq data from the small intestines and colon epithelium of three donors. This data will be clustered ('Seurat clustering & ILC phenotype analysis'); ILC expression data will be extracted and subsequently used in the 'Metabolic pathway analysis' section; it will also be combined with GSE125527 and used in the 'ILC phenotype plasticity analysis' section.

GSE125527 contains primary scRNA-seq data of immune cells taken from fifteen patients presenting with ulcerative colitis/IBD. This data will be clustered ('Seurat clustering & ILC phenotype analysis'); ILC expression data will be extracted and subsequently used in the 'Metabolic pathway analysis' section; it will also be combined with GSE185224 and used in the 'ILC phenotype plasticity analysis' section.

```{r Importing datasets}

# GSE 150050, four tissues
# Read in the counts data
counts.15 <- read.csv("C:/Users/Me/OneDrive - University of Nebraska at Omaha/Administrative/Documents/Senior Project/Data/GSE150050/STAR_raw_counts.csv", row.names=1)

# Read in the metadata
metadata.15 <- read.csv("C:/Users/Me/OneDrive - University of Nebraska at Omaha/Administrative/Documents/Senior Project/Data/GSE150050/GSE150050_metadata.csv", sep = ";", row.names=1)

# SQL script is used to narrow metadata.15 to colon samples
metadata15 <- as.data.frame(metadata.15)
metadata.15.colon <- sqldf("SELECT * FROM metadata15 WHERE TISSUE = 'COLON'")

# Metadata indicates all colon samples have distinct identifier 'GNI' in cell ID
counts15 <- as.data.frame(counts.15)
counts.15.colon <- counts15[,grepl("GNI", colnames(counts15))]

# Seurat object creation
seurat.15 <- CreateSeuratObject(counts = counts.15)

# GSE 185224, normal gut
# Read in the counts data
counts.18 <- Read10X_h5("C:/Users/Me/OneDrive - University of Nebraska at Omaha/Administrative/Documents/Senior Project/Data/GSE185224/GSE185224_Donor1_filtered_feature_bc_matrix.h5", use.names=TRUE)

# Seurat object creation from Gene Expression subset (not Antibody Capture subset)
seurat.18 <- CreateSeuratObject(counts = counts.18$`Gene Expression`)
                                
# GSE 125527, IBD gut
# Read in the PBMC counts data from directory and append
dir.12 <- "C:/Users/Me/OneDrive - University of Nebraska at Omaha/Administrative/Documents/Senior Project/Data/GSE125527/UMI/"
files.12 <- list.files(path = dir.12, pattern = ".tsv.gz", full.names = TRUE)
files.12.pbmc <- files.12[grep("pBMC",files.12)]
counts.12 <- read.csv(files.12.pbmc[1],sep="\t", row.names=1)
for(i in 2:length(files.12.pbmc)){
  counts.12b <- read.csv(files.12.pbmc[i],sep="\t", row.names=1)
  counts.12 <- rbind(counts.12, counts.12b)
}
counts.12.pbmcs <- t(counts.12)

# Read in the metadata
metadata.12 <- read.csv("C:/Users/Me/OneDrive - University of Nebraska at Omaha/Administrative/Documents/Senior Project/Data/GSE125527/GSE125527_cell_metadata.csv", row.names=1)

# SQL script used to narrow metadata.12 to PBMC samples only
metadata12 <- as.data.frame(metadata.12)
metadata.12.pbmcs <- sqldf("SELECT * FROM metadata12 WHERE tissue_assignment= 'PBMC'")

# Seurat object creation
seurat.12 <- CreateSeuratObject(counts = counts.12.pbmcs, meta.data=metadata.12.pbmcs)

```

Then, the data pre-processed for further analysis. 

```{r Converting datasets to Seurat objects}

# GSE 150050, four tissues

# QC and valid cell selection

# Normalization of data

# Feature selection

# Scaling of data

# GSE 185224, normal gut

# GSE 125527, IBD gut

```

---
Clustering and phenotype composition analysis
---

This section contains code for Seurat clustering & ILC phenotype assignment.

```{r Initial clustering and cell type assignment}

# See 'Seurat - Guided clustering tutorial'

# Cluster the cells

# Run non-linear dimensional reduction (UPAM/tSNE)

# Finding differentially expressed features (cluster biomarkers)

# Assigning cell type identity to clusters


# GSE 150050, four tissues

# GSE 185224, normal gut

# GSE 125527, IBD gut

```

This section contains code for composition analysis of phenotypes between datasets.

```{r Comparing phenotype composition}

```

---
Metabolic pathway analysis
---

This section contains code for the KEGG metabolic pathway analysis sub-section. 

```{r Querying against KEGG}

# See 'pathfindR' for KEGG analysis

# Get differentially expressed genes of ILCs by dataset (dataframe gene, LFC, p-value)

# GSE 150050, four tissues

# GSE 185224, normal gut

# GSE 125527, IBD gut

#

```

The APD3 metabolic pathway analysis sub-section is conducted externally, or a portion of the dataset imported and queried.

```{r Querying against APD3}

# GSE 150050, four tissues

# GSE 185224, normal gut

# GSE 125527, IBD gut


```

---
Plasticity analysis
---

This section contains code for the ILC phenotypic plasticity analysis. 

```{r Plasticity analysis}

# ILC phenotypic plasticity analysis
# See 'Introduction to scRNA-seq integration'

# Integrate GSE 185224 (normal gut) and GSE 125527 (IBD gut)

# Perform integrated analysis 

# Identify conserved cell type markers

# Identify differential expressed genes across conditions

```

